### PR TITLE
[FIX] website : fix authenticate archived visitor

### PR DESCRIPTION
**Pull request by dbeguin**
If a visitor logs in (and has a visitor_id in his cookie), the partner will be
linked to the visitor. If, after 1 week, the visitor tries to connect again
with a different session (or another visitor_id in cookies), the authenticate
will crash because:
* _cron_archive_visitors applies on visitor inactive since at least a week
* there can be only one visitor per partner (sql constraint)
* The visitor linked to the partner is not retrieved (because archived) and
  we try to link the partner to a new visitor.

Further than that, if the visitor is archived and the linked partner wants to
login again with a new visitor_id, we should : 
* reactivate the previous visitor,
* copy history from newest to previous one
* delete the newest one

(last two points were already done before this commit).

This PR also fixes and simplifies the time_statistics computation.
As time_connection_datetime is always set and in the depends, no need to read
values before looping, the data is already fetched in memory. We can than
use directly the value for each visitor in self.

Task ID: 2120464

### CODE REVIEW

File: addons/website/models/website_visitor.py [hunk 1]
```diff
@@ -117,13 +117,9 @@ def _compute_last_visited_page_id(self):
...
     def _compute_time_statistics(self):
-        results = self.env['website.visitor'].with_context(active_test=False).search_read([('id', 'in', self.ids)], ['id', 'last_connection_datetime'])
-        mapped_data = {result['id']: result['last_connection_datetime'] for result in results}
-
         for visitor in self:
-            last_connection_date = mapped_data[visitor.id]
-            visitor.time_since_last_action = _format_time_ago(self.env, (datetime.now() - last_connection_date))
-            visitor.is_connected = (datetime.now() - last_connection_date) < timedelta(minutes=5)
+            visitor.time_since_last_action = _format_time_ago(self.env, (datetime.now() - visitor.last_connection_datetime))
+            visitor.is_connected = (datetime.now() - visitor.last_connection_datetime) < timedelta(minutes=5)
```
Comment (awa-odoo) [inline addons/website/models/website_visitor.py]: Is this an actual fix? meaning it did NOT work before?

It does not seem related to the first issue though.
At least split in 2 commits if it's an actual fix.

File: addons/website/models/res_users.py [hunk 2]
```diff
@@ -72,10 +72,14 @@ def authenticate(cls, db, login, password, user_agent_env):
                 visitor_sudo = env['website.visitor']._get_visitor_from_request()
                 if visitor_sudo:
                     partner = env.user.partner_id
-                    partner_visitor = env['website.visitor'].sudo().search([('partner_id', '=', partner.id)])
+                    partner_visitor = env['website.visitor'].with_context(active_test=False).sudo().search([('partner_id', '=', partner.id)])
                     if partner_visitor and partner_visitor.id != visitor_sudo.id:
+                        # Link history to older Visitor and delete the newest
                         visitor_sudo.website_track_ids.write({'visitor_id': partner_visitor.id})
                         visitor_sudo.unlink()
+                        # Reactivate the older Visitor & Update visitor's Cookie (TODO: check if necessary (maybe done in _handle_webpage_dispatch)
```
Comment (awa-odoo) [inline addons/website/models/res_users.py]: 1. I don't like that TODO, can you actually check?
2. It's unclear here that visitors are archived by a cron, should be explained in the docstring AND in the commit message
3. I don't see any cookie update, the comment is a little confusing
4. I attach an image related to all points above

brain

File: addons/website/models/website_visitor.py [hunk 3]
```diff
@@ -117,13 +117,9 @@ def _compute_last_visited_page_id(self):
...
     @api.depends('last_connection_datetime')
     def _compute_time_statistics(self):
-        results = self.env['website.visitor'].with_context(active_test=False).search_read([('id', 'in', self.ids)], ['id', 'last_connection_datetime'])
-        mapped_data = {result['id']: result['last_connection_datetime'] for result in results}
-
         for visitor in self:
-            last_connection_date = mapped_data[visitor.id]
-            visitor.time_since_last_action = _format_time_ago(self.env, (datetime.now() - last_connection_date))
-            visitor.is_connected = (datetime.now() - last_connection_date) < timedelta(minutes=5)
+            visitor.time_since_last_action = _format_time_ago(self.env, (datetime.now() - visitor.last_connection_datetime))
```
Comment (awa-odoo) [inline addons/website/models/website_visitor.py]: I'm sorry to ask again but what does it "fix" exactly?
Really sounds like performance improvement / cleaning to me.

File: addons/website/models/res_users.py [hunk 4]
```diff
@@ -72,10 +72,14 @@ def authenticate(cls, db, login, password, user_agent_env):
                 visitor_sudo = env['website.visitor']._get_visitor_from_request()
                 if visitor_sudo:
                     partner = env.user.partner_id
-                    partner_visitor = env['website.visitor'].sudo().search([('partner_id', '=', partner.id)])
+                    partner_visitor = env['website.visitor'].with_context(active_test=False).sudo().search([('partner_id', '=', partner.id)])
                     if partner_visitor and partner_visitor.id != visitor_sudo.id:
+                        # Link history to older Visitor and delete the newest
                         visitor_sudo.website_track_ids.write({'visitor_id': partner_visitor.id})
                         visitor_sudo.unlink()
+                        # Reactivate the older Visitor
```
Comment (awa-odoo) [inline addons/website/models/res_users.py]: `# Reactivate the archived visitor (most likely archived by the cron for inactivity reasons)`

File: addons/website/models/res_users.py [hunk 5]
```diff
@@ -72,10 +72,14 @@ def authenticate(cls, db, login, password, user_agent_env):
                 visitor_sudo = env['website.visitor']._get_visitor_from_request()
                 if visitor_sudo:
                     partner = env.user.partner_id
-                    partner_visitor = env['website.visitor'].sudo().search([('partner_id', '=', partner.id)])
+                    partner_visitor = env['website.visitor'].with_context(active_test=False).sudo().search([('partner_id', '=', partner.id)])
                     if partner_visitor and partner_visitor.id != visitor_sudo.id:
+                        # Link history to older Visitor and delete the newest
```
Comment (awa-odoo) [inline addons/website/models/res_users.py]: Seems to me that the _get_visitor_from_request() should NOT create the visitor in that specific use case.
It's a little pointless to create a visitor to immediately unlink it, right?

Anyway, that would be for another task targeting master.

File: addons/website/models/website_visitor.py (hunk 3 above)
Comment (dbeguin) [reply addons/website/models/website_visitor.py]: Indeed, this is not a fix (anymore). There was a time where RLI (RIP) did this PR : #38457

File: addons/website/models/res_users.py (hunk 5 above)
Comment (dbeguin) [reply addons/website/models/res_users.py]: No because the visitor is already created (not in authenticate process but) in dispatch process.
In authenticate, we just retrieve the visitor from the visitor's cookie.

**Comment by dbeguin** [comment]
@awa-odoo : tests added for this particular flow.
Comments applied or answered !
Bisous!

File: addons/website/models/website_visitor.py (hunk 3 above)
Comment (awa-odoo) [reply addons/website/models/website_visitor.py]: As discussed:

Change it back to FIX and explain it prevents a stack trace when creating visitors in the backend.

File: addons/website/models/res_users.py (hunk 5 above)
Comment (awa-odoo) [reply addons/website/models/res_users.py]: Ok, use case a little pourri, but ok

File: addons/website/tests/test_website_visitor.py [hunk 6]
```diff
@@ -1,13 +1,16 @@
 # coding: utf-8
-from odoo.tests import HttpCase
+from odoo import tests
+from datetime import datetime, timedelta
 
 
-class WebsiteVisitorTests(HttpCase):
-    def test_create_visitor_on_tracked_page(self):
+@tests.tagged('dbetest')
```
Comment (awa-odoo) [inline addons/website/tests/test_website_visitor.py]: oopsy

File: addons/website/tests/test_website_visitor.py [hunk 7]
```diff
@@ -43,14 +46,54 @@ def test_create_visitor_on_tracked_page(self):
...
+    def test_long_period_inactivity(self):
+        # link visitor to partner
+        old_visitor = self.Visitor.create({
+            'lang_id': self.env.ref('base.lang_en').id,
+            'country_id': self.env.ref('base.be').id,
+            'website_id': 1,
+        })
+        partner_demo = self.env.ref('base.partner_demo')
+        old_visitor.partner_id = partner_demo.id
+        partner_demo.visitor_ids = [(6, 0, [old_visitor.id])]  # TODO DBE : To remove in Master (13.1) after visitor_ids field declaration master fix
```
Comment (awa-odoo) [inline addons/website/tests/test_website_visitor.py]: change to `[4, old_visitor.id]`

File: addons/website/tests/test_website_visitor.py [hunk 8]
```diff
@@ -43,14 +46,54 @@ def test_create_visitor_on_tracked_page(self):
...
+        self.assertEqual(partner_demo.visitor_ids.id, old_visitor.id, "The partner visitor should be set correctly.")
+
+        # archive old visitor
+        old_visitor.last_connection_datetime = datetime.now() - timedelta(days=8)
+        self.Visitor._cron_archive_visitors()
+        self.assertEqual(old_visitor.active, False, "The visitor should be archived after one week of inactivity")
+
+        # reconnect with new visitor.
+        self.url_open(self.tracked_view.url)
+        new_visitor = self.Visitor.search([('id', '!=', old_visitor.id)])
```
Comment (awa-odoo) [inline addons/website/tests/test_website_visitor.py]: Then you should limit=1 with the appropriate order (id desc I guess?)

File: addons/website/tests/test_website_visitor.py [hunk 9]
```diff
@@ -43,14 +46,54 @@ def test_create_visitor_on_tracked_page(self):
...
+        self.assertEqual(len(new_visitor), 1, "A visitor should be created after visiting a tracked view")
+        self.assertEqual(len(self.Track.search([])), 1, "A track should be created after visiting a tracked view")
```
Comment (awa-odoo) [inline addons/website/tests/test_website_visitor.py]: Can't you specify a track "for that visitor"?

File: addons/website/tests/test_website_visitor.py [hunk 10]
```diff
@@ -43,14 +46,54 @@ def test_create_visitor_on_tracked_page(self):
...
+
+        # override the get_visitor_from_request to mock that is new_visitor that authenticates
+        self.target_visitor = new_visitor
+        def get_visitor_from_request(self_mock, force_create=False):
+            if self.target_visitor:
+                return self.target_visitor
+            else:
+                return self.Visitor._create_visitor() if force_create else self.Visitor
```
Comment (awa-odoo) [inline addons/website/tests/test_website_visitor.py]: As you define `self.target_visitor` just above, can you reach the else case?

File: addons/website/tests/test_website_visitor.py [hunk 11]
```diff
@@ -43,14 +46,54 @@ def test_create_visitor_on_tracked_page(self):
...
+        self.patch(type(self.env['website.visitor']), '_get_visitor_from_request', get_visitor_from_request)
+
+        self.authenticate('demo', 'demo')
+        self.assertEqual(partner_demo.visitor_ids.id, old_visitor.id, "The partner visitor is incorrect.")
```
Comment (awa-odoo) [inline addons/website/tests/test_website_visitor.py]: "The partner visitor should be back to the 'old' visitor."

File: addons/website/tests/test_website_visitor.py [hunk 12]
```diff
@@ -43,14 +46,54 @@ def test_create_visitor_on_tracked_page(self):
...
+
+        new_visitor = self.Visitor.search([('id', '!=', old_visitor.id)])
+        self.assertEqual(len(new_visitor), 0, "The new visitor should be deleted when visitor authenticate once again.")
+        self.assertEqual(old_visitor.active, True, "The old visitor should be reactivated when visitor authenticate once again.")
```
Comment (awa-odoo) [inline addons/website/tests/test_website_visitor.py]: authenticateS

File: addons/website/tests/test_website_visitor.py (hunk 10 above)
Comment (dbeguin) [reply addons/website/tests/test_website_visitor.py]: ah indeed, oopsy (I defined it in the setup originally)

File: addons/website/tests/test_website_visitor.py (hunk 7 above)
Comment (dbeguin) [reply addons/website/tests/test_website_visitor.py]: I will remove this anyway after forward port (And I need to flush if I use [4, id]

File: addons/website/tests/test_website_visitor.py (hunk 6 above)
Comment (awa-odoo) [inline addons/website/tests/test_website_visitor.py]: DAVID! :D

File: addons/website/tests/test_website_visitor.py [hunk 13]
```diff
@@ -1,13 +1,16 @@
...
+from odoo import tests
+from datetime import datetime, timedelta
 
 
-class WebsiteVisitorTests(HttpCase):
-    def test_create_visitor_on_tracked_page(self):
+@tests.tagged('dbetest')
+class WebsiteVisitorTests(tests.HttpCase):
+    def setUp(self):
+        super().setUp()
```
Comment (awa-odoo) [inline addons/website/tests/test_website_visitor.py]: Is this a valid way to call super?

I've always seen `super(WebsiteVisitorTests, self).setUp()`

**Comment by dbeguin** [comment]
aaaand.. voilà :)

**Comment by tde-banana-odoo** [comment]
@robodoo r+ rebase-ff

**Comment by dbeguin** [comment]
@d-fence or @KangOl : This PR seems not to be forward ported yet to master. Is it normal ?

**Comment by dbeguin** [comment]
Ah, apparently, a part of this PR have been removed in the FW PR. strange..

**Comment by actonsys** [comment]
I think this error is related to this PR. 
This is happening on production mode and it's been thrown several times when different users are trying to acces the website. If reload the page, the webpage comes back to normal, but after a while, the same issue.

> 2020-02-07 05:02:44,643 31740 ERROR DB odoo.sql_db: bad query: INSERT INTO "website_track" ("id", "page_id", "url", "visit_datetime", "visitor_id") VALUES (nextval('website_track_id_seq'), 4, 'https://www.WEBSITEXXX', '2020-02-07 05:02:44.638207', 43) RETURNING id
ERROR: could not serialize access due to concurrent update
CONTEXT:  SQL statement "SELECT 1 FROM ONLY "public"."website_visitor" x WHERE "id" OPERATOR(pg_catalog.=) $1 FOR KEY SHARE OF x"

**Comment by tde-banana-odoo** [comment]
We will have a look. Thanks for pointing it out !
